### ISSUE TITLE
Cross Validation with LightGBMPruningCallback & XGBoostPruningCallback

### ISSUE THREAD

**Issue by nielsuit227** [optuna/optuna#3203]
When using the `LightGBMPruningCallback` or `XGBoostPruningCallback` with cross-validation, only the first fold is used, and the others are ignored as the step is already reported. 
Currently, `LightGBMPruningCallback` allows to use a step name (`valid_0`), which requires a new instantiation of the the pruner every fold. `XGBoostPruningCallback`, doesn't allow this at all. It simply takes the first fold. How can I do this better? 

### Environment
Optuna 2.10.0
Python 3.9.1
Windows 10 - 10.0.19041

### Error messages, stack traces, or logs

```
[I 2021-12-21 17:16:29,020] Trial 4 finished with value: 0.174946319013834 and parameters: {'max_depth': 8}. Best is trial 3 with value: 0.16761709988117218.
[0]     train-mlogloss:0.73793
[1]     train-mlogloss:0.52498
[2]     train-mlogloss:0.38364
[3]     train-mlogloss:0.28694
[4]     train-mlogloss:0.21818
[5]     train-mlogloss:0.16928
[6]     train-mlogloss:0.13318
[7]     train-mlogloss:0.10683
[8]     train-mlogloss:0.08766
[9]     train-mlogloss:0.07287
[0]     train-mlogloss:0.73298
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 0 is already reported.
  warnings.warn(
[1]     train-mlogloss:0.52036
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 1 is already reported.
  warnings.warn(
[2]     train-mlogloss:0.38232
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 2 is already reported.
  warnings.warn(
[3]     train-mlogloss:0.28857
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 3 is already reported.
  warnings.warn(
[4]     train-mlogloss:0.22250
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 4 is already reported.
  warnings.warn(
[5]     train-mlogloss:0.17487
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 5 is already reported.
  warnings.warn(
[6]     train-mlogloss:0.14021
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 6 is already reported.
  warnings.warn(
[7]     train-mlogloss:0.11478
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 7 is already reported.
  warnings.warn(
[8]     train-mlogloss:0.09549
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 8 is already reported.
  warnings.warn(
[9]     train-mlogloss:0.08109
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 9 is already reported.
  warnings.warn(
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 0 is already reported.
  warnings.warn(
[0]     train-mlogloss:0.74552
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 1 is already reported.
  warnings.warn(
[1]     train-mlogloss:0.53604
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 2 is already reported.
  warnings.warn(
[2]     train-mlogloss:0.40075
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 3 is already reported.
  warnings.warn(
[3]     train-mlogloss:0.30639
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 4 is already reported.
  warnings.warn(
[4]     train-mlogloss:0.23898
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 5 is already reported.
  warnings.warn(
[5]     train-mlogloss:0.18920
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 6 is already reported.
  warnings.warn(
[6]     train-mlogloss:0.15119
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 7 is already reported.
  warnings.warn(
[7]     train-mlogloss:0.12311
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 8 is already reported.
  warnings.warn(
[8]     train-mlogloss:0.10188
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 9 is already reported.
  warnings.warn(
[9]     train-mlogloss:0.08505
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 0 is already reported.
  warnings.warn(
[0]     train-mlogloss:0.73944
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 1 is already reported.
  warnings.warn(
[1]     train-mlogloss:0.52869
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 2 is already reported.
  warnings.warn(
[2]     train-mlogloss:0.39098
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 3 is already reported.
  warnings.warn(
[3]     train-mlogloss:0.29622
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 4 is already reported.
  warnings.warn(
[4]     train-mlogloss:0.22902
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 5 is already reported.
  warnings.warn(
[5]     train-mlogloss:0.18025
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 6 is already reported.
  warnings.warn(
[6]     train-mlogloss:0.14435
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 7 is already reported.
  warnings.warn(
[7]     train-mlogloss:0.11768
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 8 is already reported.
  warnings.warn(
[8]     train-mlogloss:0.09635
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 9 is already reported.
  warnings.warn(
[9]     train-mlogloss:0.08056
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 0 is already reported.
  warnings.warn(
[0]     train-mlogloss:0.74808
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 1 is already reported.
  warnings.warn(
[1]     train-mlogloss:0.53855
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial.py:590: UserWarning: The reported value is ignored because this `step` 2 is already reported.
  warnings.warn(
[2]     train-mlogloss:0.40068
C:\Users\AppData\Local\Programs\Python\Python39\lib\site-packages\optuna\trial\_trial
```

### Reproducible examples

```python
import optuna
from sklearn.datasets import load_iris
from sklearn.model_selection import KFold
from sklearn.metrics import log_loss
import xgboost as xgb


def objective(trial):
    x, y = load_iris(return_X_y=True)
    scores = []
    for t, v in KFold(shuffle=True).split(x, y):
        xt, xv = xgb.DMatrix(x[t], label=y[t]), xgb.DMatrix(x[v])
        params = {
            'max_depth': trial.suggest_int('max_depth', 3, 15),
            'eval_metric': 'mlogloss',
            'num_class': 3,
            'objective': 'multi:softprob'}
        model = xgb.train(params, xt, evals=[(xt, 'train')], callbacks=[optuna.integration.XGBoostPruningCallback(trial, 'train-mlogloss')])
        yp = model.predict(xv)
        scores.append(log_loss(y[v], yp))
    return sum(scores) / len(scores)


study = optuna.create_study(direction='minimize')
study.optimize(objective, n_trials=10)
```

## Note to the questioner

If you are more comfortable with [Stack Overflow], you may consider posting your question with the "optuna" tag there instead.
Alternatively, for issues that would benefit from more of an interactive session with the developers,
you may refer to the [optuna/optuna chat] on Gitter.

[Stack Overflow]: https://stackoverflow.com/questions/tagged/optuna
[optuna/optuna chat]: https://gitter.im/optuna/optuna

**Comment by nzw0301** [comment]
For LightGBM, [`LightGBMTunerCV`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.lightgbm.LightGBMTunerCV.html#optuna.integration.lightgbm.LightGBMTunerCV) might be that you would like to use. For xgboost, unfortunately, optuna does not have a similar interface to `LightGBMTunerCV`. Technically, we need to report the averaged validation metirc over the k-folds at each iteration.

**Comment by nielsuit227** [comment]
Thanks for your comment! 
`LightGBMTunerCV` is a wrapper for `LightGBM.train()`, and given we use our own wrappers, I would like to avoid using this. 
Reporting the average validation metric defeats the purpose of the pruner, as it should prune the model itself, not the trial. Whereas the average validation metric is only known at the end of the trial. 
In my opinion, we should be able to feed an argument to the callbacks, in which we specify the number of folds. This would solely suppress the warnings of multiple reports on the same step. The reported values of each fold would be used, pruning unpromising folds, which destroys the performance of the fold and therefore the trial. After all, a robust model is only as strong as its performance on the weakest fold. Do you think it's worthwhile changing this into a feature request so that something could be organised?

**Comment by nzw0301** [comment]
Hi, sorry for the delayed response. Indeed, I misunderstood the behaviour of `LightGBMTunerCV`. 

I suppose https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.OptunaSearchCV.html is doing a similar thing. So if this integration can be used with xgboost or lightgbm, it might be a solution.

**Comment by nzw0301** [comment]
You are right, lightgbm provides scikit-learn interface but it doesn't have `partial_fit`; `OptunaSearchCV` does not work. Sorry for confusing you.

**Comment by acrofales** [comment]
My current workaround is to only activate the pruner in the first fold of CV, but just wanted to point out that this is also a problem for pruning trials in Catboost. I was looking for a way to modify the callback in a way that it can access the fold, but I don't really see any way of doing this. Specifying the number of folds, or somehow indicating that a new fold has started would be effective solutions. The trial should be pruned if on any of the folds the pruner detects it is a bad trial.

**Comment by Julian-Stricker** [comment]
@acrofales Could you explain how you implemented the workaround you mentioned (only activate the pruner in the first fold of CV)? Thanks!

**Comment by acrofales** [comment]
> @acrofales Could you explain how you implemented the workaround you mentioned (only activate the pruner in the first fold of CV)? Thanks!

Sure. I implemented the CV manually (in a for loop) and only pass the callback on the first iteration:

```
splitter = StratifiedShuffleSplit(n_splits=5, test_size=0.2, random_state=1234)
for i, (train_index, validate_index) in enumerate(splitter.split(X, y)):
    ... # snipped processing data.
    if i == 0:
        clf_pipeline.fit(
                X_train,
                y_train,
                catboostclassifier__eval_set=validate_pool,
                catboostclassifier__callbacks=[pruning_callback],
        )
    else:
        clf_pipeline.fit(X_train, y_train, catboostclassifier__eval_set=evalpool)
```

clf_pipeline is an sklearn pipeline, so need to pass the catboost-specific parameters separated with an underscore. if it's just a straight up catboost classifier, you can use the normal parameter names, of course.

**Comment by carlosg-m** [comment]
> My current workaround is to only activate the pruner in the first fold of CV, but just wanted to point out that this is also a problem for pruning trials in Catboost. I was looking for a way to modify the callback in a way that it can access the fold, but I don't really see any way of doing this. Specifying the number of folds, or somehow indicating that a new fold has started would be effective solutions. The trial should be pruned if on any of the folds the pruner detects it is a bad trial.

@acrofales, do you think it makes sense to modify your solution and apply prunning to a random fold instead of a pre-defined fold?

```
scores = []
pruning_fold = np.random.randint(low=0, high=n_splits)
    for (fold_i, (train_fold, val_fold)) in enumerate(sgkf_splitter.split(......)):
        
        # train sets
        xtrain = df_features.iloc[train_fold]
        ytrain = df_label.iloc[train_fold]

        # fit model
        pruning_callback = None
        if pruning_fold == fold_i:
            pruning_callback = LightGBMPruningCallback(trial, metric='avg_precision_macro', valid_name='valid')
```
